### PR TITLE
fix(compiler): Add parentheses around C# casts to fix #1714

### LINKED ISSUE #1714: Miscompilation due to incorrect parenthetisation in C# output for casts
Dafny miscompiles `(x as Y).z` into `x.z as Y`.  Here's a concrete example:

```dafny
trait Base {}
class Derived extends Base { var n: int; constructor() { n := 0; } }

method f(b: Base) {
  if (b is Derived) {
    print((b as Derived).n);
  }
}

method Main() {
  var d := new Derived();
  f(d);
}
```

```console
$ Dafny /compile:3 casts.dfy
Dafny program verifier finished with 2 verified, 0 errors
Errors compiling program:
(1846,42): error CS1061: 'Base' does not contain a definition for 'n' and no accessible extension method 'n' accepting a first argument of type 'Base' could be found (are you missing a using directive or an assembly reference?)
```

This is the (broken) generated code:

```
    public static void f(Base b)
    {
      if (Dafny.Helpers.Let<Base, bool>(b, _is_0 => _is_0 is Derived)) {
        Dafny.Helpers.Print((Derived)(b).n);
      }
    }
```

`(Derived)(b).n` associates as `(Derived)((b).n)`, whereas Dafny expects it to associate as `((Derived)(b)).n`, hence the problem.

### PR DESCRIPTION
**Pull request by cpitclaudel**
Fixes #1714

### CODE REVIEW

File: Test/git-issues/git-issue-1714.dfy.expect [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+Dafny program verifier finished with 2 verified, 0 errors
+Running...
+
+(b as Derived).n == 0
```
Comment (fabiomadge) [inline Test/git-issues/git-issue-1714.dfy.expect:5]:
```suggestion
Dafny program verifier finished with 2 verified, 0 errors
(b as Derived).n == 0
```
This should fix the test.

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -11,7 +13,6 @@
 - fix: Regain C# 7.3 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
 - fix: The error "assertion violation" is replaced by the better wording "assertion might not hold". This indicates better that the verifier is unable to prove the assertion. (https://github.com/dafny-lang/dafny/pull/1862)
 
-
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:14]: Let's keep this for consistency?

Comment (fabiomadge) [reply RELEASE_NOTES.md:14]: Or, remove the others as well.

File: Source/Dafny/Compilers/Compiler-Csharp.cs [hunk 3]
```diff
@@ -1479,7 +1479,7 @@ protected class ClassWriter : IClassWriter {
         if (Attributes.ContainsBool(cl.Attributes, "handle", ref isHandle) && isHandle) {
           return "0";
         } else {
-          return $"({TypeName(xType, wr, udt.tok)})null";
+          return $"(({TypeName(xType, wr, udt.tok)})null)";
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-Csharp.cs:1482]: Do you expect this will ever part of a bigger expression?

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -11,7 +13,6 @@
 - fix: Regain C# 7.3 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
 - fix: The error "assertion violation" is replaced by the better wording "assertion might not hold". This indicates better that the verifier is unable to prove the assertion. (https://github.com/dafny-lang/dafny/pull/1862)
```
Comment (cpitclaudel) [inline RELEASE_NOTES.md:14]:
```suggestion
- fix: The error "assertion violation" is replaced by the better wording "assertion might not hold". This indicates better that the verifier is unable to prove the assertion. (https://github.com/dafny-lang/dafny/pull/1862)

```

File: Source/Dafny/Compilers/Compiler-Csharp.cs (hunk 3 above)
Comment (cpitclaudel) [reply Source/Dafny/Compilers/Compiler-Csharp.cs:1482]: I changed it for consistency with `return $"(({TypeName(xType, wr, udt.tok)})null)";` above

**Review by fabiomadge — APPROVED** [review]

**Pull request merged**